### PR TITLE
fix: FSDP mixed precision initializes params in fp32, not bf16/fp16

### LINKED ISSUE #21556: Fabric FSDP precision: incorrect dtype during init when using bf16-mixed
### Bug description

When using fabric FSDP bf16-mixed precision, the initialized parameters will be bf16, despite the intended semantics being maintaining the parameters in full precision i.e., fp32 (as suggested by the [doc](https://lightning.ai/docs/fabric/stable/api/fabric_args.html#precision)). This is due to the fact that the `module_init_context` uses `param_dtype`, which is bf16 ([here](https://github.com/Lightning-AI/pytorch-lightning/blob/8d86b2417d56703d5cc4f0da76acc71717d1dab7/src/lightning/fabric/plugins/precision/fsdp.py#L114-L115)).

### What version are you seeing the problem on?

master

### Reproduced in studio

_No response_

### How to reproduce the bug

```python
# minimal_fsdp_fabric_bf16_mixed.py
import torch
from torch import nn

from lightning.fabric import Fabric
from lightning.fabric.plugins.precision import FSDPPrecision
from lightning.fabric.strategies import FSDPStrategy


class TinyModel(nn.Module):
    def __init__(self, in_dim: int = 1024, hidden: int = 4096, out_dim: int = 10):
        super().__init__()
        self.net = nn.Sequential(
            nn.Linear(in_dim, hidden),
            nn.GELU(),
            nn.Linear(hidden, out_dim),
        )

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        return self.net(x)


# FSDP strategy (keep it minimal; no custom wrapping policy here)
strategy = FSDPStrategy()

# For FSDP, use the FSDP-specific precision plugin when you want "bf16-mixed"
fabric = Fabric(
    accelerator="cuda",
    devices="auto",
    strategy=strategy,
    plugins=FSDPPrecision("bf16-mixed"),
)

fabric.launch()

# ---- model initialization (stop here, as requested) ----
# fabric.init_module() ensures parameters get initialized correctly for the chosen strategy/precision
with fabric.init_module():
    model = TinyModel()

print(next(model.parameters()).dtype)
```

### Error messages and logs

torch.bfloat16

### Environment

lightning==2.6.1

### More info

_No response_

cc @ethanwharris @justusschock @lantiga

### PR DESCRIPTION
**Pull request by bhimrazy**
## What does this PR do?

Fixes #21556

When using `FSDPPrecision` with `bf16-mixed` or `16-mixed`, calling `fabric.init_module()` (or `configure_model()` in Trainer) was initializing model parameters in bf16/fp16 instead of fp32.

Mixed-precision training keeps a full-precision (fp32) copy of parameters — only the forward/backward computation is cast to lower precision via autocast. This is the standard AMP semantics (Micikevicius et al. 2018) and what PyTorch's own `MixedPrecision` docs say: *"outside forward and backward, the sharded parameters are kept in full precision."*

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--21586.org.readthedocs.build/en/21586/

### CODE REVIEW

**Comment by bhimrazy** [comment]
requires #21587

**Review by justusschock — APPROVED** [review]
